File: cdattg_web_golang/services/carnet_biblioteca.go
/**
 * Armo el listado de biblioteca: solo regulares que el líder ya aprobó.
 *
 * @author Cristian Deysdayr Jiménez
 */
package services

import (
	"errors"

	"github.com/sena/cdattg-web-golang/dto"
	"github.com/sena/cdattg-web-golang/models"
)

var errCarnetNoBiblioteca = errors.New("este carnet no es de formación regular aprobada")

// ListarBiblioteca trae fichas y personas con carnet regular validado.
func (s *carnetDigitalService) ListarBiblioteca(fichaID uint) (*dto.CarnetBibliotecaResponse, error) {
	list, err := s.solicitudRepo.FindAprobadosRegular()
	if err != nil {
		if tablaCarnetAusente(err) {
			return &dto.CarnetBibliotecaResponse{Fichas: []dto.CarnetBibliotecaFicha{}, Items: []dto.CarnetBibliotecaItem{}}, nil
		}
		return nil, err
	}
	lideres, err := s.solicitudRepo.FindNombresLiderPorFicha(fichaIDsDeSolicitudes(list))
	if err != nil {
		return nil, err
	}
	personas, err := s.solicitudRepo.FindPersonasPorIDs(personaIDsDeSolicitudes(list))
	if err != nil {
		return nil, err
	}
	out := bibliotecaDesdeSolicitudes(list, lideres, personas)
	out.Items = filtrarItemsBiblioteca(out.Items, fichaID)
	return &out, nil
}

// LeerFotoBiblioteca entrega la foto si el carnet es regular y está aprobado.
func (s *carnetDigitalService) LeerFotoBiblioteca(solicitudID uint) (*PersonaFotoArchivo, error) {
	sol, err := s.solicitudRepo.FindByID(solicitudID)
	if err != nil {
		return nil, err
	}
	if !esCarnetParaBiblioteca(sol) {
		return nil, errCarnetNoBiblioteca
	}
	return leerFotoPersona(sol.FotoPath)
}

func esCarnetParaBiblioteca(sol *models.CarnetSolicitud) bool {
	return sol != nil && sol.Estado == models.CarnetEstadoAprobado && sol.TipoFormacion == models.TipoFormacionRegular
}

File: cdattg_web_frontend/src/pages/carnets/biblioteca/carnetBiblioteca.ts
/**
 * Filtro las personas de biblioteca por la ficha que eligió el usuario.
 *
 * @author Cristian Deysdayr Jiménez
 */
import type { CarnetBibliotecaItem } from '../../../types/carnet';

/**
 * Dejo solo los de esa ficha. Si no hay ficha, dejo todos.
 * @param items personas con carnet regular aprobado
 * @param fichaId ficha elegida; 0 = todas
 * @returns personas visibles
 */
export function filtrarItemsBiblioteca(items: CarnetBibliotecaItem[], fichaId: number): CarnetBibliotecaItem[] {
  if (fichaId <= 0) {
    return items;
  }
  return items.filter((it) => it.ficha_id === fichaId);
}
